File: src/formats/vtt.ts
// WebVTT (.vtt) adapter.
//
// Format gotchas:
//  - Must start with "WEBVTT". Header text + region/style is preserved in meta.
//  - Time: "HH:MM:SS.mmm --> HH:MM:SS.mmm" (dot separator, hours optional).
//  - Cue settings after the arrow (align, position, line…) are preserved in raw.
//  - NOTE blocks and STYLE blocks are preserved verbatim in meta (round-trip).
//  - Inline timestamps "<00:00:02.000>" split text into timed tokens (karaoke).
//    These map to/from Cue.tokens.

import { newCueId, emptyDocument, type Cue, type SubtitleDocument, type SyncToken } from "../types/subtitle";
import { formatVttTime, parseClockTime } from "../utils/time";
import { sortedCues } from "./srt";

const ARROW_RE = /-->/;
const INLINE_TS_RE = /<(\d{1,2}:)?\d{1,2}:\d{1,2}\.\d{1,3}>/g;

export function parseVtt(raw: string): SubtitleDocument {
  const doc = emptyDocument("vtt");
  const text = raw.replace(/\r\n/g, "\n").replace(/^﻿/, "");
  const blocks = text.split(/\n\s*\n/);

  const preamble: string[] = [];
  for (let bi = 0; bi < blocks.length; bi++) {
    const block = blocks[bi].trim();
    if (!block) continue;

    // Header block (first one) and NOTE/STYLE/REGION blocks are kept in meta.
    if (bi === 0 && block.startsWith("WEBVTT")) {
      preamble.push(block);
      continue;
    }
    if (/^(NOTE|STYLE|REGION)\b/.test(block)) {
      preamble.push(block);
      continue;
    }

    const lines = block.split("\n");
    // Optional cue identifier line (no arrow) precedes the timing line.
    let identifier: string | undefined;
    if (lines.length && !ARROW_RE.test(lines[0])) {
      identifier = lines.shift()!.trim();
    }
    if (!lines.length || !ARROW_RE.test(lines[0])) continue;

    const timeLine = lines.shift()!;
    const arrowIdx = timeLine.search(ARROW_RE);
    const startStr = timeLine.slice(0, arrowIdx).trim();
    const rest = timeLine.slice(arrowIdx + 3).trim();
    const restParts = rest.split(/\s+/);
    const endStr = restParts.shift() ?? "";
    const settings = restParts.join(" ");

    const start = parseClockTime(startStr);
    const end = parseClockTime(endStr);
    if (start == null || end == null) continue;

    const rawText = lines.join("\n");
    const tokens = extractTokens(rawText, start, end);
    const cue: Cue = {
      id: newCueId(),
      start,
      end,
      text: stripInlineTimestamps(rawText).trim(),
    };
    if (identifier) cue.raw = { ...cue.raw, identifier };
    if (settings) cue.raw = { ...cue.raw, settings };
    if (tokens.length) cue.tokens = tokens;
    doc.cues.push(cue);
  }

  if (preamble.length) doc.meta.vttPreamble = preamble.join("\n\n");
  return doc;
}

export function serializeVtt(doc: SubtitleDocument): string {
  const header = doc.meta.vttPreamble ?? "WEBVTT";
  const cues = sortedCues(doc.cues);
  const parts = cues.map((cue) => {
    const id = cue.raw?.identifier ? `${cue.raw.identifier}\n` : "";
    const settings = cue.raw?.settings ? ` ${cue.raw.settings}` : "";
    const body = cue.tokens?.length ? renderTokens(cue) : cue.text;
    return `${id}${formatVttTime(cue.start)} --> ${formatVttTime(cue.end)}${settings}\n${body}`;
  });
  return [header, ...parts].join("\n\n") + "\n";
}

function stripInlineTimestamps(text: string): string {
  return text.replace(INLINE_TS_RE, "");
}

/** Parse "word<00:00:02.000>next" into timed tokens within [start, end]. */
function extractTokens(text: string, start: number, end: number): SyncToken[] {
  if (!INLINE_TS_RE.test(text)) return [];
  INLINE_TS_RE.lastIndex = 0;

  const tokens: SyncToken[] = [];
  let cursor = 0;
  let segStart = start;
  let match: RegExpExecArray | null;
  while ((match = INLINE_TS_RE.exec(text)) !== null) {
    const word = text.slice(cursor, match.index).trim();
    const ts = parseClockTime(match[0].slice(1, -1));
    if (word && ts != null) {
      tokens.push({ text: word, start: segStart, end: ts });
      segStart = ts;
    } else if (ts != null) {
      segStart = ts;
    }
    cursor = match.index + match[0].length;
  }
  const tail = text.slice(cursor).trim();
  if (tail) tokens.push({ text: tail, start: segStart, end });
  return tokens;
}

function renderTokens(cue: Cue): string {
  const toks = cue.tokens!;
  let out = "";
  toks.forEach((tok, i) => {
    if (i > 0) out += `<${formatVttTime(tok.start)}>`;
    out += (i > 0 ? " " : "") + tok.text;
  });
  return out;
}

File: native/Sources/GlyphlineCore/Formats/VTT.swift
// WebVTT (.vtt) adapter (ported from ../../src/formats/vtt.ts).
//
//  - Header + NOTE/STYLE/REGION blocks preserved verbatim in meta.vttPreamble.
//  - Cue identifier + settings preserved in cue.raw.
//  - Inline timestamps "<00:00:02.000>" map to/from Cue.tokens (karaoke).

import Foundation

private let inlineTSRegex = try! NSRegularExpression(pattern: #"<(\d{1,2}:)?\d{1,2}:\d{1,2}\.\d{1,3}>"#)

public func parseVtt(_ raw: String) -> SubtitleDocument {
    var doc = SubtitleDocument.empty(.vtt)
    let text = raw.replacingOccurrences(of: "\r\n", with: "\n").strippingLeadingBOM()
    let blocks = regexSplit(#"\n\s*\n"#, text)
    var preamble: [String] = []

    for (bi, rawBlock) in blocks.enumerated() {
        let block = rawBlock.trimmed()
        if block.isEmpty { continue }
        if bi == 0 && block.hasPrefix("WEBVTT") { preamble.append(block); continue }
        if regexTest(#"^(NOTE|STYLE|REGION)\b"#, block) { preamble.append(block); continue }

        var lines = block.components(separatedBy: "\n")
        var identifier: String?
        if let first = lines.first, !first.contains("-->") {
            identifier = first.trimmed()
            lines.removeFirst()
        }
        guard let firstLine = lines.first, firstLine.contains("-->") else { continue }
        let timeLine = lines.removeFirst()

        guard let arrow = timeLine.range(of: "-->") else { continue }
        let startStr = String(timeLine[timeLine.startIndex..<arrow.lowerBound]).trimmed()
        let rest = String(timeLine[arrow.upperBound...]).trimmed()
        var restParts = rest.split(whereSeparator: { $0 == " " || $0 == "\t" }).map(String.init)
        let endStr = restParts.isEmpty ? "" : restParts.removeFirst()
        let settings = restParts.joined(separator: " ")

        guard let start = parseClockTime(startStr), let end = parseClockTime(endStr) else { continue }

        let rawText = lines.joined(separator: "\n")
        let tokens = extractVttTokens(rawText, start, end)
        var rawFields: [String: String] = [:]
        if let identifier, !identifier.isEmpty { rawFields["identifier"] = identifier }
        if !settings.isEmpty { rawFields["settings"] = settings }

        doc.cues.append(Cue(
            id: newCueId(),
            start: start,
            end: end,
            text: stripInlineTimestamps(rawText).trimmed(),
            tokens: tokens.isEmpty ? nil : tokens,
            raw: rawFields.isEmpty ? nil : rawFields
        ))
    }

    if !preamble.isEmpty { doc.meta["vttPreamble"] = preamble.joined(separator: "\n\n") }
    return doc
}

public func serializeVtt(_ doc: SubtitleDocument) -> String {
    let header = doc.meta["vttPreamble"] ?? "WEBVTT"
    let parts = sortedCues(doc.cues).map { cue -> String in
        let id = cue.raw?["identifier"].map { "\($0)\n" } ?? ""
        let settings = cue.raw?["settings"].map { " \($0)" } ?? ""
        let body = (cue.tokens?.isEmpty == false) ? renderVttTokens(cue) : cue.text
        return "\(id)\(formatVttTime(cue.start)) --> \(formatVttTime(cue.end))\(settings)\n\(body)"
    }
    return ([header] + parts).joined(separator: "\n\n") + "\n"
}

private func stripInlineTimestamps(_ text: String) -> String {
    let ns = text as NSString
    return inlineTSRegex.stringByReplacingMatches(
        in: text, range: NSRange(location: 0, length: ns.length), withTemplate: ""
    )
}

private func extractVttTokens(_ text: String, _ start: Double, _ end: Double) -> [SyncToken] {
    let ns = text as NSString
    let matches = inlineTSRegex.matches(in: text, range: NSRange(location: 0, length: ns.length))
    if matches.isEmpty { return [] }
    var tokens: [SyncToken] = []
    var cursor = 0
    var segStart = start
    for m in matches {
        let word = ns.substring(with: NSRange(location: cursor, length: m.range.location - cursor)).trimmed()
        let tsStr = ns.substring(with: m.range)
        let ts = parseClockTime(String(tsStr.dropFirst().dropLast())) // strip < >
        if !word.isEmpty, let ts {
            tokens.append(SyncToken(text: word, start: segStart, end: ts))
            segStart = ts
        } else if let ts {
            segStart = ts
        }
        cursor = m.range.location + m.range.length
    }
    let tail = ns.substring(from: cursor).trimmed()
    if !tail.isEmpty { tokens.append(SyncToken(text: tail, start: segStart, end: end)) }
    return tokens
}

private func renderVttTokens(_ cue: Cue) -> String {
    guard let toks = cue.tokens else { return cue.text }
    var out = ""
    for (i, tok) in toks.enumerated() {
        if i > 0 { out += "<\(formatVttTime(tok.start))>" }
        out += (i > 0 ? " " : "") + tok.text
    }
    return out
}
